prov/sm2/src/sm2_progress.c

```diff
@@ -201,11 +202,139 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
 	return ret < 0 ? ret : 0;
 }
 
+static void sm2_do_atomic(void *src, void *dst, void *cmp,
+			  enum fi_datatype datatype, enum fi_op op, size_t cnt,
+			  uint32_t op_flags)
+{
+	char tmp_result[SM2_ATOMIC_INJECT_SIZE];
+
+	if (ofi_atomic_isswap_op(op)) {
+		ofi_atomic_swap_handler(op, datatype, dst, src, cmp, tmp_result,
+					cnt);
+	} else if (op_flags & FI_REMOTE_READ && ofi_atomic_isreadwrite_op(op)) {
+		ofi_atomic_readwrite_handler(op, datatype, dst, src, tmp_result,
+					     cnt);
+	} else if (ofi_atomic_iswrite_op(op)) {
+		ofi_atomic_write_handler(op, datatype, dst, src, cnt);
+	} else {
+		FI_WARN(&sm2_prov, FI_LOG_EP_DATA,
+			"invalid atomic operation\n");
+	}
+
+	if (op_flags & FI_REMOTE_READ)
+		memcpy(src, op == FI_ATOMIC_READ ? dst : tmp_result,
+		       cnt * ofi_datatype_size(datatype));
+}
+
+static int sm2_progress_inject_atomic(struct sm2_xfer_entry *xfer_entry,
+				      struct fi_ioc *ioc, size_t ioc_count,
+				      size_t *len, struct sm2_ep *ep)
+{
+	struct sm2_atomic_entry *atomic_entry =
+		(struct sm2_atomic_entry *) xfer_entry->user_data;
+	uint8_t *src, *comp;
+	int i;
+
+	switch (xfer_entry->hdr.op) {
+	case ofi_op_atomic_compare:
+		src = atomic_entry->atomic_data.buf;
+		comp = atomic_entry->atomic_data.comp;
+		break;
+	default:
+		src = atomic_entry->atomic_data.data;
+		comp = NULL;
+		break;
+	}
+
+	for (i = *len = 0; i < ioc_count && *len < xfer_entry->hdr.size; i++) {
+		sm2_do_atomic(&src[*len], ioc[i].addr,
+			      comp ? &comp[*len] : NULL,
+			      atomic_entry->atomic_hdr.datatype,
+			      atomic_entry->atomic_hdr.atomic_op, ioc[i].count,
+			      xfer_entry->hdr.op_flags);
+		*len += ioc[i].count *
+			ofi_datatype_size(atomic_entry->atomic_hdr.datatype);
+	}
+
+	if (*len != xfer_entry->hdr.size) {
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL, "recv truncated");
+		return -FI_ETRUNC;
+	}
+
+	return FI_SUCCESS;
+}
+
+static int sm2_progress_atomic(struct sm2_ep *ep,
+			       struct sm2_xfer_entry *xfer_entry)
+{
+	struct sm2_atomic_entry *atomic_entry =
+		(struct sm2_atomic_entry *) xfer_entry->user_data;
+	struct sm2_domain *domain = container_of(
+		ep->util_ep.domain, struct sm2_domain, util_domain);
+	struct fi_ioc ioc[SM2_IOV_LIMIT];
+	size_t i;
+	size_t ioc_count = atomic_entry->atomic_hdr.rma_ioc_count;
+	size_t total_len = 0;
+	int err = 0, ret = 0;
+	struct fi_rma_ioc *ioc_ptr;
+
+	for (i = 0; i < ioc_count; i++) {
+		ioc_ptr = &(atomic_entry->atomic_hdr.rma_ioc[i]);
+		ret = ofi_mr_verify(
+			&domain->util_domain.mr_map,
+			ioc_ptr->count *
+				ofi_datatype_size(
+					atomic_entry->atomic_hdr.datatype),
+			(uintptr_t *) &(ioc_ptr->addr), ioc_ptr->key,
+			ofi_rx_mr_reg_flags(
+				xfer_entry->hdr.op,
+				atomic_entry->atomic_hdr.atomic_op));
```

Comment:
This function got all sorts of wonky. If it's not too much trouble with clang, can you fix how it handles functions called inside parameters? Otherwise, don't really care, just will grumble about it for now.

Reply:
Lets talk more about what looking better entails... I think the issue is with the multiplication / having our line length be so short.

Reply:
How about something like

```
	size_t datatype_sz;
	uint64_t mr_flags;

	for (i = 0; i < ioc_count; i++) {
		ioc_ptr = &(atomic_entry->atomic_hdr.rma_ioc[i]);

		datatype_sz =
			ofi_datatype_size(atomic_entry->atomic_hdr.datatype);
		mr_flags = ofi_rx_mr_reg_flags(
			xfer_entry->hdr.op, atomic_entry->atomic_hdr.atomic_op);

		ret = ofi_mr_verify(&domain->util_domain.mr_map,
				    ioc_ptr->count * datatype_sz,
				    (uintptr_t *) &(ioc_ptr->addr),
				    ioc_ptr->key, mr_flags);
		if (ret)
			break;

		ioc[i].addr = (void *) ioc_ptr->addr;
		ioc[i].count = ioc_ptr->count;
	}
```